Add Descriptive MosViz Columns, other MosViz improvements

dtn5ah opened the pull request:
This PR does a couple of things:
1. Adds the Source Name, RA, and Dec of each target (via the 'OBJECT', 'OBJ_RA', and 'OBJ_Dec' fits keywords)
2. Adds a metadata parser to extract the above information
3. Adds a standard `load_data()` method to automatically call the 1d, 2d, image, and metadata parsers
4. Updates the MOSviz Example Notebook to automatically download the data and use the standard `load_data()` loader by default.

nmearl commented on jdaviz/configs/mosviz/helper.py:
```diff
@@ -5,7 +5,70 @@
+
+    _default_configuration = "mosviz"
+
+    def load_data(
+        self,
+        onedspectra,
+        twodspectra,
+        images,
+        onedspectra_label=None,
+        twodspectra_label=None,
+        images_label=None,
+    ):
```
```suggestion
    def load_data(self, onedspectra, twodspectra, images, onedspectra_label=None, 
                  twodspectra_label=None, images_label=None):
```

To be consistent with the rest of the codebase, line breaks should really only happen when necessary and where appropriate.

nmearl commented on jdaviz/configs/mosviz/helper.py:
```diff
@@ -22,8 +85,9 @@ def load_1d_spectra(self, data_obj, data_labels=None):
         """
-        self.app.load_data(data_obj, parser_reference="mosviz-spec1d-parser",
-                           data_labels=data_labels)
+        self.app.load_data(
+            data_obj, parser_reference="mosviz-spec1d-parser", data_labels=data_labels
+        )
```
```suggestion
        self.app.load_data(data_obj, parser_reference="mosviz-spec1d-parser",
                           data_labels=data_labels)
```

The original format was one less line and kept the line under 79 characters.

nmearl commented on jdaviz/configs/mosviz/helper.py:
```diff
@@ -40,8 +104,9 @@ def load_2d_spectra(self, data_obj, data_labels=None):
         """
-        self.app.load_data(data_obj, parser_reference="mosviz-spec2d-parser",
-                           data_labels=data_labels)
+        self.app.load_data(
+            data_obj, parser_reference="mosviz-spec2d-parser", data_labels=data_labels
+        )
```
```suggestion
        self.app.load_data(data_obj, parser_reference="mosviz-spec2d-parser",
                           data_labels=data_labels)
```

See above.

nmearl commented on jdaviz/configs/mosviz/helper.py:
```diff
@@ -59,8 +124,9 @@ def load_images(self, data_obj, data_labels=None):
         """
-        self.app.load_data(data_obj, parser_reference="mosviz-image-parser",
-                           data_labels=data_labels)
+        self.app.load_data(
+            data_obj, parser_reference="mosviz-image-parser", data_labels=data_labels
+        )
```
```suggestion
        self.app.load_data(data_obj, parser_reference="mosviz-image-parser",
                           data_labels=data_labels)
```

See above.

nmearl commented on jdaviz/configs/mosviz/helper.py:
```diff
@@ -73,7 +139,7 @@ def add_column(self, data, column_name=None):
             Header string to be shown in the table.
         """
-        table_data = self.app.data_collection['MOS Table']
+        table_data = self.app.data_collection["MOS Table"]
```
```suggestion
        table_data = self.app.data_collection['MOS Table']
```
This has not been "set in stone", so to speak, but the unspoken formalism so far has been, essentially, double quotes for text, especially text that will be used with string interpolation, and single quotes for anything that behaves like an identifier (e.g. `MOS Table` in this case).

It has been more-or-less up to developers to decide how to use their single or double quotes, but I think the basic rule is that we shouldn't tread on each other's chosen formalism if it doesn't conflict with some dev-team codebase-level standards.

dtn5ah replied:
Ahh oops, that's my bad, and wasn't a conscious decision. I have VScode configured to automatically format my code on save with the [black formatter](https://black.readthedocs.io/en/stable/), as that was the one we were experimenting with in Green Bank. Totally agree with your statement, I should switch that to something less aggressive

nmearl commented on jdaviz/configs/mosviz/helper.py:
```diff
@@ -86,7 +152,7 @@ def to_table(self):
             An astropy table constructed from the loaded mos data.
         """
-        table_data = self.app.data_collection['MOS Table']
+        table_data = self.app.data_collection["MOS Table"]
```
See above.

nmearl commented on jdaviz/configs/mosviz/plugins/parsers.py:
```diff
@@ -11,7 +11,7 @@
 from pathlib import Path
 
-__all__ = ['mos_spec1d_parser', 'mos_spec2d_parser', 'mos_image_parser']
+__all__ = ["mos_spec1d_parser", "mos_spec2d_parser", "mos_image_parser"]
```
See above.

nmearl commented on jdaviz/configs/mosviz/plugins/parsers.py:
```diff
@@ -71,12 +71,10 @@ def mos_spec1d_parser(app, data_obj, data_labels=None):
-        data_obj = [Spectrum1D.read(x)
-                    if _check_is_file(x) else x
-                    for x in data_obj]
+        data_obj = [Spectrum1D.read(x) if _check_is_file(x) else x for x in data_obj]
```
```suggestion
        data_obj = [Spectrum1D.read(x)
                    if _check_is_file(x) else x
                    for x in data_obj]
```

This change results in a line greater than 79 characters, and is less readable in my opinion.

nmearl commented on jdaviz/configs/mosviz/plugins/parsers.py:
```diff
@@ -139,9 +137,7 @@ def _parse_as_cube(path):
-        data_obj = [_parse_as_cube(x)
-                    if _check_is_file(x) else x
-                    for x in data_obj]
+        data_obj = [_parse_as_cube(x) if _check_is_file(x) else x for x in data_obj]
```
```suggestion
        data_obj = [_parse_as_cube(x)
                    if _check_is_file(x) else x
                    for x in data_obj]
```

See above.

nmearl commented on jdaviz/configs/mosviz/plugins/parsers.py:
```diff
@@ -174,23 +170,21 @@ def mos_image_parser(app, data_obj, data_labels=None):
-        data_obj = [_parse_as_image(x)
-                    if _check_is_file(x) else x
-                    for x in data_obj]
+        data_obj = [_parse_as_image(x) if _check_is_file(x) else x for x in data_obj]
```
```suggestion
        data_obj = [_parse_as_image(x)
                    if _check_is_file(x) else x
                    for x in data_obj]
```

See above.

nmearl commented on jdaviz/configs/mosviz/plugins/parsers.py:
```diff
@@ -200,4 +194,33 @@ def _parse_as_image(path):
         app.data_collection[data_labels[i]] = data_obj[i]
 
-    _add_to_table(app, data_labels, 'Images')
+    _add_to_table(app, data_labels, "Images")
```
```suggestion
    _add_to_table(app, data_labels, 'Images')
```
See above.

nmearl commented on jdaviz/configs/mosviz/helper.py:
```diff
@@ -5,7 +5,70 @@
+
+    def load_data(
+        self,
+        onedspectra,
```
`onedspectra` is a bit of word jumble and bit hard to parse? Maybe `spectra_1d`, or `oned_spectra`?

nmearl commented on jdaviz/configs/mosviz/plugins/parsers.py:
```diff
@@ -200,4 +194,33 @@ def _parse_as_image(path):
+    ra = [x.header.get("OBJ_RA", float("nan")) for x in data_obj]
+    dec = [x.header.get("OBJ_DEC", float("nan")) for x in data_obj]
+    names = [x.header.get("OBJECT", "Unspecified Target") for x in data_obj]
+
```
```suggestion

[x.close() for x in data_obj]

```

Be sure to close the file objects when done or else we might cause some file handling errors. 😁

javerbukh commented on jdaviz/configs/mosviz/plugins/parsers.py:
```diff
@@ -29,17 +29,17 @@ def _add_to_table(app, data, comp_label):
         app.data_collection.append(table_data)
 
-        mos_table = app.data_collection['MOS Table']
+        mos_table = app.data_collection["MOS Table"]
```
I believe your IDE undid the previous changes that Nick mentioned (" for text and ' for things that behave like identifiers).

rosteen commented:
I pulled from this PR as a starting point for my NIRISS loader branch - any sense of how close this is to being merged? I'll have to rebase and resolve any differences there once this is done before merging my work.

dtn5ah commented:
Ok I finally got time to undo all the changes my autoformatter did. I think I got all of them. @javerbukh I also fixed your x.close error you reported; it was a side effect of some of @nmearl's changes I just blindly accepted without thinking about how I extracted the header info. Thanks for catching that!
